src/js/links.js

```diff
@@ -0,0 +1,251 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/**
+ * Singleton that provides access to all links contained in the grid (including
+ * the ones that don't fit on the grid). A link is a plain object that looks
+ * like this:
+ *
+ * {
+ *   url: "http://www.mozilla.org/",
+ *   title: "Mozilla",
+ *   frecency: 1337,
+ *   lastVisitDate: 1394678824766431,
+ * }
+ */
+ /*globals gPinnedLinks, gBlockedLinks, ProviderManager, DirectoryLinksProvider, async*/
+
+"use strict";
+
+// The maximum number of links Links.getLinks will return.
+const LINKS_GET_LINKS_LIMIT = 100;
+
+const Links = {
+
+  maxNumLinks: LINKS_GET_LINKS_LIMIT,
+
+  /**
+   * A mapping from each provider to an object { sortedLinks, siteMap, linkMap }.
+   * sortedLinks is the cached, sorted array of links for the provider.
+   * siteMap is a mapping from base domains to URL count associated with the domain.
+   *         The count does not include blocked URLs. siteMap is used to look up a
+   *         user's top sites that can be targeted with a suggested tile.
+   * linkMap is a Map from link URLs to link objects.
+   */
+  _providers: new Map(),
+
+  /**
+   * A list of objects that are observing links updates.
+   */
+  _observers: [],
+
+  /**
+   * The properties of link objects used to sort them.
+   */
+  _sortProperties: [
+    "frecency",
+    "lastVisitDate",
+    "url",
+  ],
+
+  /**
+   * Compares two links.
+   *
+   * @param {Link} aLink1 The first link.
+   * @param {Link} aLink2 The second link.
+   * @return {Integer} A negative number if aLink1 is ordered before aLink2, zero if
+   *         aLink1 and aLink2 have the same ordering, or a positive number if
+   *         aLink1 is ordered after aLink2.
+   *
+   * NOTE: compareLinks's this object is bound to Links below.
+   */
+  compareLinks(aLink1, aLink2) {
+    for (var prop of this._sortProperties) {
+      if (!(prop in aLink1) || !(prop in aLink2)) {
+        throw new Error("Comparable link missing required property: " + prop);
+      }
+    }
+    return aLink2.frecency - aLink1.frecency ||
+           aLink2.lastVisitDate - aLink1.lastVisitDate ||
+           aLink1.url.localeCompare(aLink2.url);
+  },
+
+  /**
+   * Merges the cached lists of links from all providers whose lists are cached.
+   *
+   * @return {Array} The merged list.
+   */
+  _getMergedProviderLinks() {
```

Comment:
Apologies, I had a bit of a hard time understanding this one, but it appears all you want is:

 * get all providers
 * get only the ones with sortedLinks
 * reduce them to a single array of links
 * sort the array
 * limit the size of the array to 100

Is that correct? if so, it could simply be:

```JS
    var linkLists = Array.from(this._providers.values())
      .filter(provider => provider.sortedLinks)
      .reduce((arr, provider) => provider.sortedLinks.slice(), [])
      .sort(this.compareLinks)
      .slice(0, this.maxNumLinks);
```